### PR TITLE
OFI:  add Intel Level Zero (aka ZE) to OFI MTL/BTL

### PR DESCRIPTION
**Pull request by hppritcha**
This commits adds support for registering Intel accelerator memory backed buffers with an OFI provider.

Related to #11763

### CODE REVIEW

**Review by edgargabriel — APPROVED** [review]
Looks good to me.

**Comment by hppritcha** [comment]
@wenduwan could you review when you have a chance?

**Comment by wenduwan** [comment]
@hppritcha Sure I will take a look

**Comment by wenduwan** [comment]
Have we raise the concern to libfabric community, especially to intel so the ZE hmem module could return the correct support status?

File: ompi/mca/mtl/ofi/mtl_ofi_component.c [hunk 1]
```diff
@@ -251,6 +251,15 @@ ompi_mtl_ofi_component_register(void)
...
                                     &ompi_mtl_ofi.num_ofi_contexts);
 
+    ompi_mtl_ofi.disable_hmem = false;
+    mca_base_component_var_register(&mca_mtl_ofi_component.super.mtl_version,
+                                    "disable_hmem",
+                                    "Disable HMEM usage",
+                                    MCA_BASE_VAR_TYPE_BOOL, NULL, 0, 0,
+                                    OPAL_INFO_LVL_3,
+                                    MCA_BASE_VAR_SCOPE_READONLY,
+                                    &ompi_mtl_ofi.disable_hmem);
```
Comment (wenduwan) [inline ompi/mca/mtl/ofi/mtl_ofi_component.c:261]: I'm poking the libfabric community if there is a way to disable HMEM in libfabric, so that we don't have to do this at MPI level(across different implementations).

But regardless I think this MCA parameter should be in its own commit.

Comment (edgargabriel) [reply ompi/mca/mtl/ofi/mtl_ofi_component.c:261]: my understanding of how this parameter is used was that Open MPI does not request HMEM support from libfabric. That is a different than libfabric not having been compiled with HMEM (or disabling it if it has been compiled with HMEM)

Comment (shefty) [reply ompi/mca/mtl/ofi/mtl_ofi_component.c:261]: For libfabric, you can try setting FI_HMEM_DISABLE_P2P=true.  It looks like that has the effect of disabling HMEM from the verbs provider.

Comment (wenduwan) [reply ompi/mca/mtl/ofi/mtl_ofi_component.c:261]: @edgargabriel I see. Actually you brought up a point... can we achieve the same result, i.e. Open MPI does not request HMEM support from libfabric, by turning off/disabling accelerator? Or is HMEM requested regardless of accelerator?

My understanding is that, if it's related to libfabric specifically it might be better to address it in-situ within libfabric. This could benefit Open MPI, MPICH, etc.

Comment (edgargabriel) [reply ompi/mca/mtl/ofi/mtl_ofi_component.c:261]: @wenduwan I am not sure. We could introduce the logic that if the accelerator component is the null component, we do not request HMEM support. On the hand, this might be an issue with delayed initialization of accelerator components. In addition, I am aware of valid use-cases with UCX where accelerator support from Open MPI is not required if UCX has accelerator support (e.g. no derived datatypes etc.), not entirely sure whether similar scenario could be envisioned with libfabric as well.

Comment (hppritcha) [reply ompi/mca/mtl/ofi/mtl_ofi_component.c:261]: I will see if the CXI provider honors the FI_HMEM_DISABLE_P2P env variable.  Hopefully it does.  
The problem is not in the provider itself, per se,  but rather that the driver associated with the CXI provider requires a kernel service that isn't available on SLES15 till SP4.

Comment (hppritcha) [reply ompi/mca/mtl/ofi/mtl_ofi_component.c:261]: hmmm doesn't appear that the CXI provider is heeding this environment variable.

**Comment by hppritcha** [comment]
I'll split the mca parameter additions out into a separate PR.  Its unfortunate that the OFI env variable isn't being honored by the provider in question.

**Comment by hppritcha** [comment]
new MCA parameters removed from this PR.

**Comment by hppritcha** [comment]
uh oh, looks like there's a bug in UCX - at least looking at the ompi_NVIDIA CI

**Review by wenduwan — APPROVED** [review]
